src/logic_utils/canonical_forms.cpp

```diff
@@ -0,0 +1,157 @@
+#include <algorithm>
+#include <iostream>
+#include <stack>
+#include <string>
+#include <vector>
+using namespace std;
+#include "truth_table_generator.cpp"
+
+// Function to find unique variables in the expression
+vector<Token> unique(vector<Token> expression) {
+  vector<Token> answer;
+  for (int i = 0; i < expression.size(); i++) {
+    if (expression[i].type == TokenType::VAR) {
+      // Check if the variable already exists in the answer vector
+      for (int j = 0; j < i; j++) {
+        if (expression[i].value == expression[j].value) {
+          break;
+        }
+      }
+      // Add the unique variable to the answer vector
+      answer.push_back(expression[i]);
+    }
+  }
+  return answer;
+}
+
+// Function to generate the canonical SOP (Sum of Products) form
+vector<Token> canonical_forms_sop(string expression) {
+  vector<vector<bool>> truth_table = generateTruthTable(expression);
+  vector<Token> tokenized_string = tokenize(expression);
+  vector<Token> uniqueVariables = unique(tokenized_string);

```

Comment:
You don't need to call `generateTruthTable` or `tokenize` as both the TT and the tokens will be given to the function as inputs by the main caller, so that it is more decoupled.